Metropolis-Hastings (+ readme files)

brunoguindani opened the pull request:
I've implemented versions of basic logistic regression, both Metropolis-Hastings and Langevin-adjusted, as agreed. Calculations are [here](https://www.overleaf.com/project/60252ac8a4eba11e480403a4) as usual.
I'm running it as a standalone file (see the notebook for the command) because I would not need to compile the whole library, and they are temporary files anyway.

The problem is that I get really weird acceptance ratios in both cases. In particular, this is caused by what I call "ratio2", that is the second term of the ratio involving product over data points. The first and last term seem fine, but ratio2 always seems to be either >1 for all data, or <1 for all data. This leads to either gigantic or infinitely small acceptance ratios.

I don't think MH is meant to work with what are basically 0-or-1 probabilities. But I don't understand where the error is.
The estimated coefficients don't come close to the true ones either, and to me this is further proof that there is an error somewhere. But computations seem fine, too. I'm kinda drawing a blank right now.

(I also pushed CONTRIBUTORS.md and correlated changes as promised.)

mberaha commented:
I'm sorry I don't have time to go bug hunting at the moment... I wrote this code very quickly but it seems to work to me

```
#include <gtest/gtest.h>
#include <Eigen/Dense>
#include <stan/math/prim.hpp>
#include "src/utils/rng.h"

double sigmoid(double x) { return 1.0 / (1.0 + std::exp(-x)); }

double log_like(Eigen::VectorXd y, Eigen::MatrixXd x, Eigen::VectorXd alpha) {
  double out = 0.0;
  for (int i = 0; i < y.size(); i++) {
    double tmp = sigmoid(x.row(i) * alpha);
    out += y(i) * std::log(sigmoid(tmp)) +
           (1 - y(i)) * std::log(1 - sigmoid(tmp));
  }
  return out;
}

Eigen::VectorXd mh_step(Eigen::VectorXd y, Eigen::MatrixXd x,
                        Eigen::VectorXd curr_alpha, double var) {
  int dim = curr_alpha.size();
  Eigen::VectorXd out;
  // Hardcoding prior
  Eigen::VectorXd prior_mean = Eigen::VectorXd::Zero(dim);
  Eigen::MatrixXd prior_var = Eigen::MatrixXd::Identity(dim, dim) * 5;

  // propose new state
  Eigen::MatrixXd prop_covar = Eigen::MatrixXd::Identity(dim, dim) * var;
  Eigen::VectorXd prop_alpha = stan::math::multi_normal_rng(
      curr_alpha, prop_covar, bayesmix::Rng::Instance().get());

  double lik_rate = log_like(y, x, prop_alpha) - log_like(y, x, curr_alpha);
  double prior_rate =
      stan::math::multi_normal_lpdf(prop_alpha, prior_mean, prior_var) -
      stan::math::multi_normal_lpdf(curr_alpha, prior_mean, prior_var);

  double acc_rate = lik_rate + prior_rate;

  double unif = stan::math::uniform_rng(0, 1, bayesmix::Rng::Instance().get());
  if (unif < std::exp(acc_rate)) {
    std::cout << "accepted!, acc_rate: " << std::exp(acc_rate)
              << ", lik_rate: " << lik_rate << ", prior_rate: " << prior_rate
              << std::endl;
    std::cout << "prop_alpha: " << prop_alpha.transpose() << std::endl;
    out = prop_alpha;
  } else {
    out = curr_alpha;
  }
  return out;
}

TEST(mh, 1) {
  int ndata = 500;
  Eigen::MatrixXd X = Eigen::MatrixXd::Random(ndata, 2);
  Eigen::VectorXd alpha_true(2);
  alpha_true << -10, 10;

  Eigen::VectorXd y(ndata);
  for (int i = 0; i < ndata; i++) {
    double prob = sigmoid(X.row(i) * alpha_true);
    y(i) = stan::math::bernoulli_rng(prob, bayesmix::Rng::Instance().get());
  }
  std::cout << "y: " << y.transpose() << std::endl;
  std::cout << "log_like: " << log_like(y, X, alpha_true) << std::endl;

  Eigen::VectorXd alpha(2);
  alpha << 0, 0;
  for (int j = 0; j < 1000; j++) {
    alpha = mh_step(y, X, alpha, 0.5);
  }
  std::cout << "alpha = " << alpha.transpose() << std::endl;
  std::cout << "log_like: " << log_like(y, X, alpha) << std::endl;
}
```

brunoguindani commented:
Don't you worry, I figured, you had told me. I'll try and merge with your code, then if everything else fails I'll switch to the thesis for a bit.

mberaha commented:
I also tried MALA and seems to work fine:

```
Eigen::VectorXd grad(Eigen::VectorXd y, Eigen::MatrixXd x,
                     Eigen::VectorXd alpha, Eigen::MatrixXd Lambda) {
  Eigen::VectorXd out = -Lambda * alpha;
  for (int i = 0; i < y.size(); i++) {
    out += (y(i) - sigmoid(x.row(i) * alpha)) * x.row(i).transpose();
  }

  return out;
}

Eigen::VectorXd mala_step(Eigen::VectorXd y, Eigen::MatrixXd x,
                          Eigen::VectorXd curr_alpha, double stepsize) {
  int dim = curr_alpha.size();
  Eigen::VectorXd out;
  // Hardcoding prior
  Eigen::VectorXd prior_mean = Eigen::VectorXd::Zero(dim);
  Eigen::MatrixXd prior_var = Eigen::MatrixXd::Identity(dim, dim) * 5;
  Eigen::MatrixXd prior_prec = prior_var.inverse();

  // propose new state
  Eigen::VectorXd prop_mean =
      curr_alpha + stepsize * grad(y, x, curr_alpha, prior_prec);
  Eigen::MatrixXd prop_covar =
      Eigen::MatrixXd::Identity(dim, dim).array() * std::sqrt(2 * stepsize);
  Eigen::VectorXd prop_alpha = stan::math::multi_normal_rng(
      prop_mean, prop_covar, bayesmix::Rng::Instance().get());

  double lik_rate = log_like(y, x, prop_alpha) - log_like(y, x, curr_alpha);
  double prior_rate =
      stan::math::multi_normal_lpdf(prop_alpha, prior_mean, prior_var) -
      stan::math::multi_normal_lpdf(curr_alpha, prior_mean, prior_var);

  Eigen::VectorXd reverse_prop_mean =
      prop_alpha + stepsize * grad(y, x, prop_alpha, prior_prec);
  double prop_rate =
      stan::math::multi_normal_lpdf(prop_alpha, prop_mean, prop_covar) -
      stan::math::multi_normal_lpdf(curr_alpha, reverse_prop_mean, prop_covar);

  double acc_rate = lik_rate + prior_rate - prop_rate;

  double unif = stan::math::uniform_rng(0, 1, bayesmix::Rng::Instance().get());
  std::cout << "acc_rate: " << std::exp(acc_rate) << ", lik_rate: " << lik_rate
            << ", prior_rate: " << prior_rate << ", prop_rate: " << prop_rate
            << std::endl;
  std::cout << "prop_alpha: " << prop_alpha.transpose() << std::endl;
  if (unif < std::exp(acc_rate)) {
    out = prop_alpha;
  } else {
    out = curr_alpha;
  }
  return out;
}

TEST(mh, 2) {
  int ndata = 500;
  Eigen::MatrixXd X = Eigen::MatrixXd::Random(ndata, 2);
  Eigen::VectorXd alpha_true(2);
  alpha_true << -10, 10;

  Eigen::VectorXd y(ndata);
  for (int i = 0; i < ndata; i++) {
    double prob = sigmoid(X.row(i) * alpha_true);
    y(i) = stan::math::bernoulli_rng(prob, bayesmix::Rng::Instance().get());
  }
  std::cout << "y: " << y.transpose() << std::endl;
  std::cout << "log_like: " << log_like(y, X, alpha_true) << std::endl;

  Eigen::VectorXd alpha(2);
  alpha << 0, 0;
  for (int j = 0; j < 1000; j++) {
    alpha = mala_step(y, X, alpha, 0.05);
  }
  std::cout << "alpha = " << alpha.transpose() << std::endl;
  std::cout << "log_like: " << log_like(y, X, alpha) << std::endl;
}
```

brunoguindani commented:
Now it works, both with my original gaussian covariates and your uniform ones. I had implemented the evaluation of the acceptance ratio by using the final results of my computations by hand, but after seeing you directly using log-likelihoods and such, I switched to that much simpler approach. I guess there must have been some computation errors, or some important constants I forgot about. I was also not using log-values, so that could be another source of errors.

So, which one performs the best? The Langevin method is truly wondrous. It has reached almost accurate state values by the 10th (!) iteration, while regular Metropolis-Hastings takes about 80. Not a surprise, I guess.
M.A.L.A. it is, then!

brunoguindani commented:
As for the structure of the logit class, or at least the part involving updates of alpha, I was thinking about something similar to what I wrote in these draft files. In short, I used separate methods for all kinds of lpdf's as well as `draw_proposal()` and utility functions.
Also, since the Durante-Rigon algorithm leverages regular Bayesian regression, I think we should be able to use `LinRegUniHierarchy` for representing the (beta, tau) pair. As mentioned in the paper, their pseudo-code uses conditionally conjugate, separate priors for the sake of concise notation, but one may also use a normal-gamma joint prior. Therefore, we should be able to merge both beta and tau update steps into good old ~sample_given_data()~ `sample_full_cond()` using `LinRegUniHierarchy` objects.

If you agree, I shall start writing the proper class.

mberaha commented:
> So, which one performs the best? The Langevin method is truly wondrous. It has reached almost accurate state values by the 10th (!) iteration, while regular Metropolis-Hastings takes about 80. Not a surprise, I guess.
> M.A.L.A. it is, then!

That's usually my take as well, also gradient evaluation should not be that intensive, so let's go with MALA

mberaha commented:
> As for the structure of the logit class, or at least the part involving updates of alpha, I was thinking about something similar to what I wrote in these draft files. In short, I used separate methods for all kinds of lpdf's as well as `draw_proposal()` and utility functions.

Yes we should start as simple as possible. Just have a `LogitStickBreakMixing` class that handles things by itself. We can later think of generalizing MALA/MH classes if needed by other parts of the code

> Also, since the Durante-Rigon algorithm leverages regular Bayesian regression, I think we should be able to use `LinRegUniHierarchy` for representing the (beta, tau) pair. As mentioned in the paper, their pseudo-code uses conditionally conjugate, separate priors for the sake of concise notation, but one may also use a normal-gamma joint prior. Therefore, we should be able to merge both beta and tau update steps into good old ~sample_given_data()~ `sample_full_cond()` using `LinRegUniHierarchy` objects.

That's fine for me. I don't really remember what they were doing with the Hierarchy part, what I was most interested in is the `Mixing` part. Then we can plug-in any hierarchy that we want, also non-dependent ones

brunoguindani commented:
I obviously need to create a new `ConditionalAlgorithm` derived object for this. What should I name it? I mean, will we be able to reuse it for other algorithms? If so, `LogitSBAlgorithm` is not a good name.

mberaha commented:
I think the right name to it should be `BlockedGibbsAlgorithm`, since it consists of the same steps of Ishwaran's and James Blocked Gibbs (http://people.ee.duke.edu/~lcarin/Yuting3.3.06.pdf)

mberaha commented on src/mixings/conditional_mixing.h:
```diff
@@ -0,0 +1,32 @@
+
+#include "base_mixing.h"
+
+class ConditionalMixing : public BaseMixing {
```
I like the idea of having a `ConditionalMixing`.  I think that instead of calling `mass_xxx_cluster()`, conditional algorithms will simply call a `get_weights(covariate)`.

mberaha commented on src/mixings/conditional_mixing.h:
```diff
@@ -0,0 +1,32 @@
+                               std::shared_ptr<AbstractHierarchy> hier,
+                               const Eigen::RowVectorXd &covariate =
+                                   Eigen::RowVectorXd(0)) const override {
+    throw std::bad_function_call(
```
I guess that having to throw errors in derived classes is a poor design choice.

The cleanest design I can think of is to have:
```
class BaseMixing {
 public:
   virtual ~BaseMixing() {}
};

class MarginalMixing : virtual public BaseMixing {
 public:
   virtual double mass_existing_cluster(...);
   virtual double mass_new_cluster(...);
};

class ConditionalMixing: virtual public BaseMixing {
 public:
   virtual Eigen::VectorXd get_weights(...);
};
```

Then, in `MarginalAlgorithm` we store an `std::shared_ptr<MarginalMixing>` while in `ConditionalAlgorithm` we store a `ConditionalMixing`.
To keep a uniform API across all algorithms, we can define `BaseAlgorithm::set_mixing(const std::shared_ptr<BaseMixing> mixing_)` as virutal, and then `MarginalAlgorithm` we can write
```
MarginalAlgorithm::set_mixing(std::shared_ptr<BaseMixing> mixing_) {
  mixing = std::dynamic_pointer_cast<MarginalMicing>(mixing_);
}
```
same goes for ConditionalAlgorithm.

In this way, we can also have the "diamond of death", so that some mixings can be both marginal and conditional (for instance the finite-dimensional dirichlet, or the finite dimensional dirichlet with a random number of atoms, is both)

brunoguindani replied:
Sounds great!

brunoguindani replied:
Upon second thoughts, we actually do a lot of stuff with `mixing` in the base algorithm, such as checking it is non-null, using it in `get_state_as_proto()`, etc. More in general, it might not be a good idea not to have it in the base class if it is part of the API, albeit in different forms depending on the type of algorithm. We could create some type of internal getter/wrapper in both intermediate classes (marginal and conditional) that performs a dynamic cast to the appropriate type, similarly to what we were doing with Hierarchies before the refactoring. Is there any easier solution?

mberaha replied:
That's a good point!
We can actually have a CRTP also for the Algorithms if that makes life simpler, or just go with having some 'casting' methods in conditional and marginal algorithms. Final possibility is to have both a pointer to BaseMixing and a pointer to MarginalMixing, that point to the same object and use whichever pointer we need

brunoguindani replied:
> Final possibility is to have both a pointer to BaseMixing and a pointer to MarginalMixing, that point to the same object and use whichever pointer we need

This is an excellent idea! We can try to apply CRTP later if we think it would improve things. For now I'd stick to this one.

mberaha replied:
Great!

brunoguindani commented on src/mixings/logit_sb_mixing.cc:
```diff
@@ -0,0 +1,115 @@
+  }
+}
+
+void LogitSBMixing::update_state(
+    const std::vector<std::shared_ptr<AbstractHierarchy>> &unique_values,
+    const unsigned int n) {
+  // Langevin-Adjusted Metropolis-Hastings step
+  auto &rng = bayesmix::Rng::Instance().get();
+  Eigen::VectorXd state_c = state.regression_coeffs;
+  auto priorcast = cast_prior();
+  Eigen::VectorXd prior_mean = bayesmix::to_eigen(
+      priorcast->normal_prior().mean());
+  double prop_var = priorcast->proposal_var();
+  double step = priorcast->step_size();
+  // Rebuild allocations vector
+  std::vector<unsigned int> allocations;  // TODO
```
The Durante-Rigon update mandates that we know whether all data points belong or not to a previous cluster than the one currently considered. Basically, we need the whole allocations vector. We can easily rebuild it from `unique_values` which is passed as argument, but this would take time which is basically wasted since we already have `BaseAlgorithm::allocations`. Should we also pass it as argument?

mberaha replied:
I would go the easy way and just pass the allocations. We can store a pointer to the covariates during initialization, as you said those are fixed

mberaha commented on src/algorithms/conditional_algorithm.h:
```diff
@@ -2,19 +2,34 @@
 class ConditionalAlgorithm : public BaseAlgorithm {
+ protected:
+  //! Weights of clusters
+  Eigen::VectorXd weights;
```
Why do we need this member in the Algorithm class? I think we should always refer to the Mixing to get the weights

brunoguindani replied:
Right, it's just leftovers from old scraped ideas. This class is still basically unimplemented. I will remove this.

mberaha commented on src/mixings/logit_sb_mixing.h:
```diff
@@ -0,0 +1,77 @@
+  bool is_dependent() const override { return true; }
+
+  //!
+  void initialize(const unsigned int n_clust = 1) override;
```
I’m not happy with this design choice.. you’re basically saying that nclust here has the same role of nclust in the Dirichlet Mixing, while in the latter case that is truly only used for initialization, here nclust is a crucial parameter. I think it deserves a separate setter / to be in the constructor. the user must be explicit in setting this parameter, it should not be done by the Algorithm

mberaha replied:
Even better, this should be part of the Prior proto!

brunoguindani replied:
It is indeed set by the user. At line 72 of `base_algorithm.cc` we have `mixing->initialize(init_num_clusters);`. Remember that `init_num_clusters` is one of arguments of `run.cc`.

mberaha replied:
I see your point, but still `init_num_cluster` is not the right parameter name.
For the DPM, PYM that is entirely fine, it's up to the algorithm to select the initial number of cluster, then the posterior will converge to its stationary distribution and that's fine.
For LSB and truncated stick-breaking prior, the `n_clust` (which should be better called `n_components` or `n_atoms`) is part of the model specification. For me, the fact that the algorithm can change this parameter, it would be just like if the algorithm could change the base measure/prior in a mixture model, or even change the hierarchy type, during initialization.

brunoguindani replied:
`init_num_clusters` is given by the user. The algorithm just assigns it to the mixture through its initialize method, and has no control over it in the current version of the code, nor does it select anything (except interpreting the zero -- an impossible value -- as "one datum per cluster"). It's really no different than using any other setter. If this is an out-of-principle argument, I can create a setter that does the exact same thing.

You may argue `init_num_clusters` is an inappropriate name in this case, and I would agree. It was the correct name for previously-implemented marginal algorithms.

But if we were to put it into a prior object, we would have the absurdity of two potentially different input parameters that refer to the exact same propoerty and which must be the same at all times without any way to enforce this. If what you're suggesting is removing `init_num_clusters` altogether from the main arguments, it can be done, but it may be a lot of work.

mberaha replied:
My point is about modularity, the `num_clusters` parameter is defining the model as much as the choice of hierarchy, is not any less important than that. Yet, it is absurd to think that the choice of the hierarchy passes through the algorithm...
So I would say to forget about the main at the moment, we're building a library, not an executable. We could also have multiple main files if that makes life easier... 

Right now I just want to get the modularity structure right, so that we can replicate that on the Python side!!

brunoguindani replied:
> the `num_clusters` parameter is defining the model as much as the choice of hierarchy, is not any less important than that

You're right, so much so that we have a unique setter `set_initial_clusters()` that sets both the hierarchy type and the number of clusters at the same time. If the algorithm setting the hierarchy does not defy modularity, neither does the algorithm setting the number of clusters. It just places to the appropriate space what it was given by the user.

If you're still not happy with the current way of setting this value, feel free to suggest one.

mberaha replied:
You're missing my point: right now we have two phases:

1- create the objects that define the statistical model (likelihood + prior)
2- pass these objects to the algorithm in charge of doing MCMC

Now, while in the DPM/PYM the `n_clust` parameter belongs to the algorithm, in the LSB this parameter belongs to the statistical model. Moreover, an `LSB` instance without a num_clusters / num_components parameter should not be allowed.
Hence, since we're using protos to define the statistical model, in my opinion it makes sense to move this parameter to the proto defining the LSB prior.

If you're problem is the compatibility with the `BaseAlgorithm::set_initial_clusters`, it is sufficient to override it in the `BlockedGibbAlgorithm` and print out a warning that the parameter will be ignored.

brunoguindani replied:
Ok, fair enough. I'll get to work.

mberaha replied:
I think the cleanest solution is to have a `AlgorithmParams` proto message, which collects all the possible params, no need for fancy oneofs here, just one large list of possible parameters would suffice in my opinion.

brunoguindani replied:
Deal! Maybe I will work on this in a separate future PR.

mberaha commented on src/mixings/base_mixing.h:
```diff
@@ -12,19 +12,14 @@
+  //!
   std::shared_ptr<google::protobuf::Message> prior;
+  //!
+  Eigen::MatrixXd *covariates_ptr;
```
Why not use a constant reference? It should be safer imo

brunoguindani replied:
Beacuse you said "pointer" and I mindlessly did just that 🙃. But you're absolutely correct.
By the way, should its name be `covariates_ptr` or just `covariates`? I thought I would make clear that it is not the actual matrix.

brunoguindani replied:
Oh wait, I just realized using a const ref is impossible since it needs to receive a value as soon as it is created. But a pointer to a const matrix will do just fine.

mberaha replied:
damn those const references are sneaky, good catch! pointer to const object looks great!

mberaha commented on src/mixings/logit_sb_mixing.cc:
```diff
@@ -0,0 +1,154 @@
+  for (int h = 0; h < num_clusters; h++) {
+    eta(h) = covariate.dot(state.regression_coeffs.col(h));
+    weights(h) = sigmoid(eta(h));
+    for (int k = 0; k < h - 1; k++) {
```
I’m sure there is a better way to do it avoiding all the nested for loops on k. Probably just precompute all the etas, then store the cumulative products of (1-eta_h)...

mberaha commented on src/mixings/logit_sb_mixing.cc:
```diff
@@ -0,0 +1,154 @@
+    Eigen::VectorXd prop_mean =
+        state_c +
+        step * grad_log_full_cond(state_c, is_curr_clus, is_prev_clus);
+    auto prop_covar = prop_var * Eigen::MatrixXd::Identity(dim, dim);
```
This is not a proper MALA step, since in MALA the step size controls also the covariance of the proposal

brunoguindani replied:
Right, you have to use the square root just like you did in your snippet, I always forget about it.

mberaha commented:
Thanks Bruno for your effort! I guess that before implementing the BlockedGibbs algorithm, some preliminary tests can be performed on the LSB prior (either formal tests or qualitative), similarly to the main file you had for the MetropolisHastings class.
Once that's settled, BlockedGibbs all the way + simple simulation study :)

brunoguindani commented:
Thank you. I like your plan. I'll start implementing tests as soon as I finish addressing your remarks.

brunoguindani commented:
So the mixing _works_ as intended as far as reading/writing/updating goes, as evidenced by the first tests. But I don't know how I can perform a _statistical_ test without an actual algorithm. The old `metropolis` example files had no need of real allocation values unlike this class. For now I'm just passing fixed arbitrary/random stuff, but I'm sure that doesn't mean anything. Should we proceed directly to building BlockedGibbs?

mberaha commented:
> So the mixing _works_ as intended as far as reading/writing/updating goes, as evidenced by the first tests. But I don't know how I can perform a _statistical_ test without an actual algorithm. The old `metropolis` example files had no need of real allocation values unlike this class. For now I'm just passing fixed arbitrary/random stuff, but I'm sure that doesn't mean anything. Should we proceed directly to building BlockedGibbs?

As you wish. If it was me, I would just simulate a dataset of allocations and run the MCMC update for the LSB by passing the allocations, but if you want to go straight for the BlockedGibbs go ahead, it might be more difficult to debug...

For the test I have in mind, there are several possibilities. Two that are fairly easy are:
1. For observations i=1, ..., n generate the cluster label c_i ~ Uniform({1, 2, 3}), if c_i = 1, generate x_i ~ N([-5, -5], I), if c_i= 2, generate x_i ~ N([5, -5], I), if c_i = 2, generate x_i ~ N([0, 5], I) or whatever distribution you prefer as long as they are separated. Then pass to the LSB update_state method the allocations and an empty list of hiearchies, since I don't think they will be used...
2. Let alpha_1 = [0, -5], alpha_2 = [0, 5], alpha_3 = [5, 0]. For observations i=1, ... n generate x_i ~ N([0, 0], 5* I) and generate c_i with weights given by the LSB mechanism with the given alphas.

For both approaches you can also generate a test set using the same mechanism and check the predictive probability assigned to p(c* = c_true | x*) is sufficiently high
For the second approach you can also check the posterior distributions of the alphas.

brunoguindani commented:
It turns out that my question was quite dumb. Everything you said for option 1 is already there, I just need to generate covariates from different distributions instead of just using Eigen Random for the whole thing. Be right back soon...

brunoguindani commented:
I've done some debugging and there's definitely something weird going on with the computation of the gradient (lines 49-62 of the LogitSB class), but I can't quite point out what it is. I've checked [my calculations](https://www.overleaf.com/project/60252ac8a4eba11e480403a4) (see last paragraph) over and over, and they look correct to me. Basically the gradient gets disproportionately small or large with each added point, and drags the proposal mean with it. Because of these large values, the sigmoids computed in `log_like()` are either almost 0 or almost 1, resulting in `nan` when computing the logarithms.

Maybe the parameters I set are just wrong? I have no clue.

mberaha commented:
Ok I will look into this, what about using a simple MH? Does the algorithm work in that case?

brunoguindani commented:
Thank you. That would be (since using step=0 is impossible due to the variance) replacing `step` with 0 in `prop_mean = state_c + step * grad(...)`? Unfortunately, it does not seem to be the case. State values are kinda fluctuating around zero, which incidentally is the prior mean. True values are (2,2), (4,4), (8,8). At least we don't get `nan`s...

brunoguindani reviewed (commented):
I like your changes. I was thinking of storing the current/last acceptance ratio, but this is even better.

Before proceeding further, there's a few TODOs to discuss which I have highlighted below.

Also, should we merge this PR and add BlockedGibbs in a new one? The master branch would work correctly, it would just have a 'blank' BlockedGibbs.

mberaha commented on src/includes.h:
```diff
@@ -1,12 +1,14 @@
 #include "algorithms/neal8_algorithm.h"
 #include "collectors/file_collector.h"
 #include "collectors/memory_collector.h"
+// TODO LinRegUni is missing yet everything works... is this even necessary?
```
I mean, it's not harmful to have this... You only need the "load_xxx" stuff since c++ includes are recursive

brunoguindani replied:
Right, they come from the load files. Anyways, I will keep this updated then.

mberaha commented on src/mixings/logit_sb_mixing.cc:
```diff
@@ -0,0 +1,161 @@
+  double step = priorcast->step_size();
+  double prop_var = std::sqrt(2.0 * step);
+  // Loop over clusters, but last alpha is ignored
+  for (int h = 0; h < unique_values.size(); h++) {  // TODO -1?
```
We shouldn't have the last values, because nu_H = 1. Tis is because otherwise the weights won't sum to 1.
So there should be H-1 values of alpha.

mberaha commented on src/mixings/logit_sb_mixing.cc:
```diff
@@ -0,0 +1,161 @@
+
+Eigen::VectorXd LogitSBMixing::get_weights(
+    const Eigen::VectorXd &covariate /*= Eigen::VectorXd(0)*/) const {
+  // TODO design choice: no check on covariate
```
sure it's fine with me

mberaha commented on src/mixings/logit_sb_mixing.cc:
```diff
@@ -0,0 +1,161 @@
+    throw std::invalid_argument("Mixing prior was not provided");
+  }
+  auto priorcast = cast_prior();
+  num_clusters = priorcast->num_clusters();
```
this parameter should better be called `num_components` (also in the prior), since the number of clusters is not `H` but the cardinality of the set {h: exists c_i = h} which in general is smaller than `H`.

mberaha commented on proto/mixing_prior.proto:
```diff
@@ -29,3 +29,11 @@ message PYPrior {
+  oneof coeff {
+    MultiNormalDistribution normal_prior = 1;
+  }
+  double step_size = 2;
```
I'm not sure this belongs here, but ok for the moment...

mberaha commented on src/mixings/logit_sb_mixing.cc:
```diff
@@ -0,0 +1,164 @@
+    if (allocations[i] >= clust) {
+      bool is_curr_clus = (allocations[i] == clust);
+      double prob = sigmoid(covariates_ptr->row(i).dot(alpha));
+      grad += (is_curr_clus - prob) * covariates_ptr->row(i);
```
I think it's safer to explicitly cast the `is_curr_clus` to `double`, it looks kind of scary right now

brunoguindani replied:
I tested this and it works as intended. Since prob is a double, it correctly calls the `operator-` with two doubles. But I can change bool to double if you prefer.

mberaha commented on src/mixings/logit_sb_mixing.cc:
```diff
@@ -0,0 +1,164 @@
+  return grad;
+}
+
+double LogitSBMixing::log_like(const Eigen::VectorXd &alpha,
```
is this really a `log_like`? I would just have a `full_cond_lpdf` that returns what this is returning + the prior evaluation 

The A/R ratio becomes `full_cond_ratio - prop_ratio` where `full_cond_ratio = full_cond_lpdf(state_prop, h, allocations) - full_cond_lpdf(state_c, h, allocations)`

brunoguindani replied:
Good idea, I'll change it, we don't really need a separate likelihood member.
Speaking of which, I think this is indeed a likelihood if you treat the nu_ih as data. Which they kind of are, if you think of their relation to alpha as a generalized linear model. Is it? (I'm asking just for theoretical understanding)

mberaha replied:
"Everything is a likelihood if you interpret the parameters hard enough" (M. Ghandi)

brunoguindani commented:
This should be ready. Shall I merge?

mberaha commented:
> This should be ready. Shall I merge?

Sure go ahead
